daemon/containerd/image_push.go

```diff
@@ -85,10 +85,58 @@ func (i *ImageService) PushImage(ctx context.Context, sourceRef reference.Named,
 		}
 	}
 
-	return i.pushRef(ctx, sourceRef, metaHeaders, authConfig, out)
+	return i.pushRef(ctx, sourceRef, platform, metaHeaders, authConfig, out)
 }
 
-func (i *ImageService) pushRef(ctx context.Context, targetRef reference.Named, metaHeaders map[string][]string, authConfig *registry.AuthConfig, out progress.Output) (retErr error) {
+func (i *ImageService) getPushTarget(ctx context.Context, targetRef reference.Named, platform *ocispec.Platform) (ocispec.Descriptor, error) {
+	img, err := i.images.Get(ctx, targetRef.String())
+	if cerrdefs.IsNotFound(err) {
+		return ocispec.Descriptor{}, errdefs.NotFound(fmt.Errorf("tag does not exist: %s", reference.FamiliarString(targetRef)))
+	}
+
+	im, err := i.getImageManifestForPlatform(ctx, img, platform)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+
+	return im.Target(), nil
+}
+
+func (i *ImageService) getImageManifestForPlatform(ctx context.Context, img containerdimages.Image, platform *ocispec.Platform) (*ImageManifest, error) {
+	if platform == nil {
+		return i.NewImageManifest(ctx, img, img.Target)
+	}
+
+	pm := platforms.OnlyStrict(*platform)
+	var result *ImageManifest
+	err := i.walkReachableImageManifests(ctx, img, func(im *ImageManifest) error {
+		if im.IsAttestation() {
+			return nil
+		}
+
+		platform, err := im.ImagePlatform(ctx)
+		if err != nil {
+			return err
+		}
+
+		if !pm.Match(platform) {
+			return nil
+		}
+
+		result = im
+		return nil
+	})
```

Comment:
Looking at this and I think we want to use https://pkg.go.dev/github.com/containerd/containerd/images#LimitManifests somewhere in there since we only want one manifest

Reply:
Handled with check if result has already been set.